### PR TITLE
Upgrade to Julia 1.6

### PR DESCRIPTION
**Pull request by ali-ramadhan**
Just updating the `Manifest.toml` and the Buildkite Julia version number to see if everything passes. I've been using 1.6 fine on my laptop so this should test 1.6 + GPUs.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Lol so the GPU unit tests segfault when testing field setting.

**Comment by ali-ramadhan** [comment]
Everything looks good on the CPU but the GPU unit tests segfault when testing field setting (tried debugging but can't figure out why). All the other GPU tests pass although CI seems much slower for GPU tests (~3x slower?).

Could be related to segfault in https://github.com/CliMA/ClimateMachine.jl/pull/2146 ? @charleskawczynski @jakebolewski were you able to figure out why it was segfaulting?

**Comment by ali-ramadhan** [comment]
Okay so the segfault was because we were trying to set a field using an `Int128` or `UInt128`.

It's not an important test so I removed it but it's a little weird that it just started failing since the test has been in since Oceananigans.jl v0.1.0...

**Review by francispoulin — APPROVED** [review]

**Comment by charleskawczynski** [comment]
> All the other GPU tests pass although CI seems much slower for GPU tests (~3x slower?).

😞

> Could be related to segfault in [CliMA/ClimateMachine.jl#2146](https://github.com/CliMA/ClimateMachine.jl/pull/2146) ? @charleskawczynski @jakebolewski were you able to figure out why it was segfaulting?

Not yet, I'm going to try looking into it. @jakebolewski suggested first upgrading some packages first-- so I'm doing that now.

**Comment by jakebolewski** [comment]
hmm good to know about possible performance issues, we should be on the lookout.  I suspect that (one of the) segfaults is an OOB error that got masked by `@inbounds`, running that now locally.

**Comment by ali-ramadhan** [comment]
I opened an issue about the `Int128`/`UInt128` thing (https://github.com/JuliaGPU/CUDA.jl/issues/793) but will revisit this PR later to look into the performance regression.

**Comment by ali-ramadhan** [comment]
Hmmm a lot of failures due to CUDA scalar `getindex` operations even though we explicitly set `CUDA.allowscalar(true)` in `runtests.jl`...

We could take this opportunity to get rid of all scalar operations in the tests and just use `CUDA.@allowscalar` where it's needed.

Maybe new CUDA scalar operations are hurting performance and that's why GPU CI has slowed down?

**Comment by navidcy** [comment]
@ali-ramadhan, I found the same issue in GeophysicalFlows.jl
I think (but I'm not sure) that there was some change on how `CUDA.@allowscalar` works. For example, previously you could define a new array like:

```julia
CUDA.@allowscalar newarray = [i*j for i=1:3, j=1:4]
```
but now that's not possible! Instead you need to do

```julia
newarray = CUDA.@allowscalar [i*j for i=1:3, j=1:4]
```

Perhaps the GPU-connoisseurs might have some more insight on this? @maleadt?

**Comment by maleadt** [comment]
That probably shouldn't have changed, can you file an issue on CUDA.jl/GPUArrays.jl? I'll have a look next week. The only change to `@allowscalar` that comes to mind is task/thread-safety, which does come at a certain performance cost (it now does a TLS lookup instead of a simple pointer check, but the cost of that should be negligible compared to the subsequent memory transfer).

**Comment by navidcy** [comment]
> That probably shouldn't have changed, can you file an issue on CUDA.jl/GPUArrays.jl? I'll have a look next week. The only change to `@allowscalar` that comes to mind is task/thread-safety, which does come at a certain performance cost (it now does a TLS lookup instead of a simple pointer check, but the cost of that should be negligible compared to the subsequent memory transfer).

Seems I can't reproduce the supposed error so, sorry, my bad... Something else must have been the issue. 😔

**Comment by navidcy** [comment]
I fixed the doctests.

However, there is an issue with `Base.typename()` which seems to have changed behavior in Julia v1.6; see https://github.com/JuliaLang/julia/issues/40474

We should modify our `show()` overloads to convert output from, e.g., `typename(CPU)` to `CPU`.

**Comment by navidcy** [comment]
I suggest changing
https://github.com/CliMA/Oceananigans.jl/blob/8f98ddb678f39d86e2aae7afa678142a118808ca/src/Models/IncompressibleModels/show_incompressible_model.jl#L6
to 

```julia
print(io, "IncompressibleModel{"*string(Base.nameof(A))*", $(eltype(model.grid))}",
```

**Comment by ali-ramadhan** [comment]
Ah good point, thanks for looking into this @navidcy! Definitely my fault for using `Base.typename` when I knew it was an internal unexported function.

Will revisit this PR soon now that docs aren't getting stuck on Buildkite + Julia 1.5 anymore.

**Comment by navidcy** [comment]
A question that comes about is that after this PR do we test *only* on v1.6? If so, how are we making sure that code works smooth on v1.5? Should we test on both? Should we bump the julia compat requirement in the `Project.toml` file to `^1.6`?

**Comment by navidcy** [comment]
now at least all the CPU tests pass...

note how `CUDA.@allowscalar` had to be changed in https://github.com/CliMA/Oceananigans.jl/pull/1514/commits/8313b7d3e0524d5891611e2b7431793b8d6cc68b

**Comment by ali-ramadhan** [comment]
Thanks for working on this @navidcy! I'll try to `git merge master` which should fix some tests and then I'll look at fixing the other failures.

> A question that comes about is that after this PR: do we test *only* on v1.6? If so, how are we making sure that code works smooth on v1.5? Should we test on both? Should we bump the julia compat requirement in the `Project.toml` file to `^1.6`?

Ideally yeah we would test on both 1.5 and 1.6 but I don't think we have the CI resources to do so :(

We could set up some GitHub Actions pipelines to test Mac and Windows with 1.5 and only run a subset of the tests so it doesn't massively slow down CI. It wouldn't be comprehensive but we'd get more coverage.

Not sure if we'll have to only support Julia 1.6+ only soon, but that would save us from having to test on 1.5.

We might not want to jump ship to 1.6 too soon though. I experienced longer compilation times when using Julia 1.6 in developing PR #1522 so I had to switch back to 1.5. Maybe I just had a weird setup though, I should investigate more carefully. I could be the only one who's had issues...

Might be good to beta test 1.6 a little bit before switching completely.

**Comment by navidcy** [comment]
I believe it’s getting there. Still a few things to sort out. Will try to deal with them tomorrow.

**Comment by navidcy** [comment]
I'm wondering whether the errors in the `gpu simulation tests` group are related to Tullio.jl. Seems like only `Tullio#master` mentions/uses `CUDAKernels`...

https://github.com/mcabbott/Tullio.jl/commit/d3c4fde63c37ebae31b6d13413531a3a9241d59c

**Comment by navidcy** [comment]
@ali-ramadhan, all tests pass except those related with Tullio.jl I think!

I don't understand why CUDA.@allowscalar creates issues... But it does apparently. Am I wrong?

**Comment by navidcy** [comment]
@ali-ramadhan, also the gpu cubed-sphere tests don't pass. Any ideas why?

**Comment by navidcy** [comment]
Getting there!

**Comment by ali-ramadhan** [comment]
Yeah only a few more fixes!

CUDA.jl 3.0 only supports Julia 1.6+ and the next release of KernelAbstractions.jl should switch to CUDA.jl 3.0 so if this PR is open for a bit longer we might as well bump Oceananigans.jl to only support Julia 1.6 as well.

We could also potentially leave behind the weird Julia 1.5 hanging issue (e.g. #1601)...

X-Ref: https://github.com/JuliaGPU/KernelAbstractions.jl/pull/242

**Comment by navidcy** [comment]
CUDAKernels don't support 1.6 yet (or at least that was the case 2 days ago...)

**Comment by ali-ramadhan** [comment]
Yeah it should happen in https://github.com/JuliaGPU/KernelAbstractions.jl/pull/242 which looks like it might get merged soon maybe?

**Comment by navidcy** [comment]
@ali-ramadhan I'll try up-ing CUDA to ^3? (we can always revert the commit..)

**Comment by tomchor** [comment]
So, I need to use `sort` on `CuArray`s to calculate [available potential energy](https://github.com/CliMA/Oceananigans.jl/issues/1297) but I just realized that it's only available starting at CUDA version 2.6, which support Julia 1.6 only.

Can you guys give an ETA of long until the Julia 1.6 migration is complete? Looks like it's relatively close but depending on the answer I might need to change my research plans for now.

**Comment by navidcy** [comment]
I'm working on it continuously... :)
As soon as all tests pass we can discuss. We are almost there @tomchor I feel... only something Tullio-related was not passing the previous time. Now I updated Tullio and trying again. Hold on to your chair!

**Comment by glwagner** [comment]
> So, I need to use `sort` on `CuArray`s to calculate [available potential energy](https://github.com/CliMA/Oceananigans.jl/issues/1297) but I just realized that it's only available starting at CUDA version 2.6, which support Julia 1.6 only.

Just to add --- as a first step, you might try first copying your data to the CPU and sorting there. If you don't have to do it too often, you won't suffer much of a penalty.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 env:
-  JULIA_VERSION: "1.6.0"
+  JULIA_VERSION: "1.6.1"
```
Comment (glwagner) [inline .buildkite/pipeline.yml:2]: Don't stop won't stop upgrading

**Comment by navidcy** [comment]
I think now only some output writer test errors are only pending. Will try to sort them out today.

**Comment by navidcy** [comment]
This went astray... It takes so long to test that always there is some new feature coming in and then conflicts, etc... :(

**Comment by ali-ramadhan** [comment]
On the CPU it seems like things are failing with `FATAL ERROR: Symbol "__nv_llabs"not found`. I should be able to reduce it down to a minimal working example (guessing it's a CUDA.jl issue).

GPU tests don't want to run (CUDA driver might be too old on Sverdrup?) but GPU regression tests passed on Tartarus so things might be fine there.

**Comment by glwagner** [comment]
```
(base) glwagner@sverdrup:~$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Sat_Aug_25_21:08:01_CDT_2018
Cuda compilation tools, release 10.0, V10.0.130
```

What do we need?

**Comment by ali-ramadhan** [comment]
> What do we need?

Maybe just CUDA v10.1+? Before erroring, all the builds seem to print this warning:

```
Warning: This version of CUDA.jl only supports NVIDIA drivers for CUDA 10.1 or higher (yours is for CUDA 10.0.0)
```

https://buildkite.com/clima/oceananigans/builds/2396#4a537d8f-9863-489f-bb11-2520a00b83de/16-233

I feel like I've seen this warning on Buildkite before, but maybe now CUDA.jl means it.

**Comment by glwagner** [comment]
> FATAL ERROR: Symbol "__nv_llabs"not found. I should be able to reduce it down to a minimal working example (guessing it's a CUDA.jl issue).

I have seen this error when working with `KernelAbstractions` without any CUDA dependency.

**Comment by ali-ramadhan** [comment]
Finally all tests pass 🎉 Thanks @navidcy and @vchuravy for all your help!

@glwagner Let me know when it would be a good time to merge this PR and tag a new release.

Ran the incompressible model benchmarks and in general it seems that Julia 1.6 allocates more memory and is a bit slower on the CPU but a bit faster on the GPU.

# Quick benchmark

## Julia 1.6

```
Oceananigans v0.57.2
Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, cascadelake)
  GPU: TITAN V
```

```
                                            Incompressible model benchmarks
┌───────────────┬─────────────┬─────┬────────────┬────────────┬────────────┬────────────┬──────────┬────────┬─────────┐
│ Architectures │ Float_types │  Ns │        min │     median │       mean │        max │   memory │ allocs │ samples │
├───────────────┼─────────────┼─────┼────────────┼────────────┼────────────┼────────────┼──────────┼────────┼─────────┤
│           CPU │     Float64 │  32 │   4.996 ms │   5.047 ms │   5.113 ms │   5.770 ms │ 1.77 MiB │   2301 │      10 │
│           CPU │     Float64 │  64 │  34.951 ms │  35.967 ms │  36.414 ms │  41.417 ms │ 1.77 MiB │   2301 │      10 │
│           CPU │     Float64 │ 128 │ 301.074 ms │ 301.964 ms │ 302.498 ms │ 307.989 ms │ 1.77 MiB │   2301 │      10 │
│           CPU │     Float64 │ 256 │    2.894 s │    2.895 s │    2.895 s │    2.896 s │ 1.77 MiB │   2301 │       2 │
│           GPU │     Float64 │  32 │   2.859 ms │   2.923 ms │   3.025 ms │   3.987 ms │ 2.80 MiB │   6914 │      10 │
│           GPU │     Float64 │  64 │   2.912 ms │   3.101 ms │   3.308 ms │   5.368 ms │ 2.78 MiB │   6993 │      10 │
│           GPU │     Float64 │ 128 │   4.894 ms │   5.019 ms │   5.360 ms │   8.565 ms │ 2.80 MiB │   8667 │      10 │
│           GPU │     Float64 │ 256 │  33.569 ms │  36.266 ms │  36.029 ms │  36.883 ms │ 3.24 MiB │  37307 │      10 │
└───────────────┴─────────────┴─────┴────────────┴────────────┴────────────┴────────────┴──────────┴────────┴─────────┘
```

## Julia 1.5

```
Oceananigans v0.57.2
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, cascadelake)
  GPU: TITAN V
```

```
                                             Incompressible model benchmarks
┌───────────────┬─────────────┬─────┬────────────┬────────────┬────────────┬────────────┬─────────────┬────────┬─────────┐
│ Architectures │ Float_types │  Ns │        min │     median │       mean │        max │      memory │ allocs │ samples │
├───────────────┼─────────────┼─────┼────────────┼────────────┼────────────┼────────────┼─────────────┼────────┼─────────┤
│           CPU │     Float64 │  32 │   4.905 ms │   5.116 ms │   5.237 ms │   6.262 ms │  360.44 KiB │   2142 │      10 │
│           CPU │     Float64 │  64 │  34.327 ms │  36.419 ms │  36.188 ms │  37.507 ms │  360.44 KiB │   2142 │      10 │
│           CPU │     Float64 │ 128 │ 297.567 ms │ 299.404 ms │ 299.856 ms │ 303.081 ms │  360.44 KiB │   2142 │      10 │
│           CPU │     Float64 │ 256 │    2.786 s │    2.786 s │    2.786 s │    2.787 s │  360.44 KiB │   2142 │       2 │
│           GPU │     Float64 │  32 │   2.729 ms │   2.781 ms │   2.885 ms │   3.717 ms │  839.83 KiB │   7132 │      10 │
│           GPU │     Float64 │  64 │   2.894 ms │   3.055 ms │   3.136 ms │   4.046 ms │  879.45 KiB │   7124 │      10 │
│           GPU │     Float64 │ 128 │   4.107 ms │   4.836 ms │   4.890 ms │   6.086 ms │  964.73 KiB │   7142 │      10 │
│           GPU │     Float64 │ 256 │  21.667 ms │  35.642 ms │  34.245 ms │  35.746 ms │    1.11 MiB │   7134 │      10 │
└───────────────┴─────────────┴─────┴────────────┴────────────┴────────────┴────────────┴─────────────┴────────┴─────────┘
```

**Comment by navidcy** [comment]
OMG

**Review by navidcy — APPROVED** [review]
I approve so much

**Comment by tomchor** [comment]
Should we be concerned about the GPU memory allocations? They're roughly 3x larger for 1.6 which is a pretty big difference! Especially considering the size limitations on GPUs.

**Comment by navidcy** [comment]
> Thanks @ali-ramadhan! That's amazing work.
> 
> Quick quick question: Should we be concerned about the GPU memory allocations? They're roughly 3x larger for 1.6 which is a pretty big difference! Especially considering the size limitations on GPUs.

Actually that's a good point. I didn't notice this. I was so overwhelmed with the `All checks have passed` that I couldn't see straight.

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -29,7 +29,7 @@ steps:
       - "$SVERDRUP_HOME/julia-$JULIA_VERSION/bin/julia -O0 --color=yes --project -e 'using Pkg; Pkg.test()'"
     agents:
       queue: Oceananigans
-      architecture: GPU
+      architecture: CPU
```
Comment (navidcy) [inline .buildkite/pipeline.yml:32]: @ali-ramadhan, why this became `CPU`? Is this some leftover from debugging?

Comment (ali-ramadhan) [reply .buildkite/pipeline.yml]: Ah good catch! I forgot to switch back to using Sverdrup. Just added a commit to switch back.

**Comment by ali-ramadhan** [comment]
> Quick quick question: Should we be concerned about the GPU memory allocations? They're roughly 3x larger for 1.6 which is a pretty big difference! Especially considering the size limitations on GPUs.

Ah so in the benchmarks those are just CPU memory allocations since BenchmarkTools.jl doesn't measure GPU allocations. `CUDA.@time` can measure GPU allocations but I haven't used it much on Oceananigans.

I don't think it's a cause for worry but it might be good to do some profiling at some point to figure out where the extra memory allocations are coming from.

Interestingly the benchmarks suggest that GPU models are actually a bit faster now :eyes:

**Comment by glwagner** [comment]
I'm not sure --- but it's possible that some GPU utilities not under Oceananigans.jl control incur memory allocations. I don't think we have much GPU-specific code in our codebase (except for pressure solvers...) Definitely a good thing to keep tabs on and open issues in the relevant packages if it affects the performance of our code.

**Pull request merged**